opencog/atomese/atom_types/NewLink.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * asmoses/opencog/atomese/atom_types/NewLink.h
+ *
+ * Copyright (C) 2019 Bitseat Tadesse
+ *
+ * All Rights Reserved
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef ATOMSPACE_NEWLINK_H
+#define ATOMSPACE_NEWLINK_H
+
+#include <opencog/atoms/core/FunctionLink.h>
+#include <opencog/atoms/core/ScopeLink.h>
+#include <opencog/atoms/core/Quotation.h>
+#include <opencog/atoms/atom_types/atom_types.h>
+
+namespace opencog {
+    class NewLink : public FunctionLink {
```

Comment:
Initially I thought it should be called `NewLinkLink`, cause it's a link that creates a new link. On the other hand, it could be a generic function that creates new atom type, link or node. So for instance if the `TypeNode` provided as first argument is a node, with no subsequent arguments, then it would create a new node with a unique name, on the other hand if the `TypeNode` is a link, then it would creates a link. So I think it would actually work. Let me know what you think, @kasimebrahim? Others?

Reply:
I agree. Generalizing it a good idea.